### PR TITLE
Migrating Csharp validator from AutoRest

### PR DESCRIPTION
**Pull request by dsgouda**
This still needs to be tested as a plugin in AutoRest. 
- Moved the Csharp validator under this repo
- Reconfigured the gulp and travis scripts to build and test the dotnet validator
- ensured dotnet validator tests are passing as expected

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,5 +3,15 @@ dist: trusty
...
+install:
+  # node packages
+  - npm install
+  # dotnet
+  - sudo sh -c 'echo "deb [arch=amd64] https://apt-mo.trafficmanager.net/repos/dotnet/ trusty main" > /etc/apt/sources.list.d/dotnetdev.list' 
+  - sudo apt-key adv --keyserver apt-mo.trafficmanager.net --recv-keys 417A0893
+  - sudo apt-get update
+  - sudo apt-get install dotnet-dev-1.0.0-rc4-004769 -y
+  # gulp
+  - npm install gulp
```
Comment (vishrutshah) [inline .travis.yml]: why install gulp twice? 

https://travis-ci.org/Azure/azure-openapi-validator/builds/248082097#L529
https://travis-ci.org/Azure/azure-openapi-validator/builds/248082097#L1073

File: gulpfile.js [hunk 2]
```diff
@@ -25,6 +27,50 @@ gulp.task('test', ['build'], function () {
...
+
+gulp.task('restore/dotnet', ['clean/dotnet'], function () {
+    console.log('Running dotnet restore...');
+    return gulp.src('src/dotnet/**/*.csproj')
+        .pipe(restore());
+});
+
+gulp.task('build/dotnet', ['restore/dotnet'], function () {
+    console.log('Running dotnet build...');
+    return gulp.src('src/dotnet/**/*.csproj')
```
Comment (vishrutshah) [inline gulpfile.js]: curious: Do we have a solution file to work on instead of each csproj?

File: package.json [hunk 3]
```diff
@@ -32,6 +32,7 @@
     "gulp-run": "1.7.1",
     "gulp-clean": "0.3.2",
     "gulp-mocha": "4.3.1",
+    "gulp-dotnet-cli": "0.4.0",
```
Comment (vishrutshah) [inline package.json]: 👍 👍

File: src/dotnet/AutoRest/JsonRpc/CallerResponse.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
```
Comment (vishrutshah) [inline src/dotnet/AutoRest/JsonRpc/CallerResponse.cs]: Different license?

File: src/dotnet/AutoRest/JsonRpc/Connection.cs [hunk 5]
```diff
@@ -0,0 +1,383 @@
+using AutoRest.JsonRpc;
```
Comment (vishrutshah) [inline src/dotnet/AutoRest/JsonRpc/Connection.cs]: license header please

File: .travis.yml (hunk 1 above)
Comment (dsgouda) [inline .travis.yml]: I had some trouble with gulp yesterday hence the 2 installs, will try removing this.

File: gulpfile.js (hunk 2 above)
Comment (dsgouda) [inline gulpfile.js]: Yes we do, but restore works on the csprojs and I thought we could keep this consistent

File: src/dotnet/AutoRest.Core/Properties/Resources.resx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  <resheader name="reader">
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="AutoRestCore" xml:space="preserve">
+    <value>AutoRest Core {0}</value>
+  </data>
+  <data name="ConfigurationKnownPlugins" xml:space="preserve">
```
Comment (olydis) [inline src/dotnet/AutoRest.Core/Properties/Resources.resx]: this file can be cleaned up significantly 😃

Comment (dsgouda) [inline src/dotnet/AutoRest.Core/Properties/Resources.resx]: Aye, noted.

Comment (olydis) [inline src/dotnet/AutoRest.Core/Properties/Resources.resx]: I'll make a quick cleanup run, found more stuff...

Comment (dsgouda) [inline src/dotnet/AutoRest.Core/Properties/Resources.resx]: Feel free to point it out, I don't mind taking it on

**Comment by dsgouda** [comment]
@vishrutshah I am open to it, let me know when and we can gather other team members too for a quick go-over

**Comment by vishrutshah** [comment]
@dsgouda Awesome. Let's do it tomorrow then!

**Comment by dsgouda** [comment]
Pending:
Point to all the moved directories from AutoRest repo
Copy over the signing job details from AutoRest

**Comment by dsgouda** [comment]
Stuff copied from AutoRest Solution projects
[AutoRest](https://github.com/Azure/autorest/tree/master/src/core/AutoRest)
Plugin interface, JsonRPC interface constructs, main entry point to the plugin

[AutoRest.Core](https://github.com/Azure/autorest/tree/master/src/core/AutoRest.Core)
Parser, Logger and base types 

[AutoRest.Swagger](https://github.com/Azure/autorest/tree/master/src/modeler/AutoRest.Swagger)
Model representation with all the assigned attributes and renamed as OpenAPI.Validator

[AutoRest.Swagger.Tests](https://github.com/Azure/autorest/tree/master/src/modeler/AutoRest.Swagger.Tests)
All validation related tests renamed as OpenAPI.Validator.Tests

**Comment by dsgouda** [comment]
@olydis Is there a reason to prefer 2 separate packages for the typescript and dotnet validators?

**Comment by olydis** [comment]
yes: 1 package = 1 extension = 1 process to run (that assumption is good for the health of people working on the core)
So: unless you could provide one "entry point" command that would somehow exposes both validators (perfectly possible, e.g. you could spawn the dotnet process from the TypeScript validator), it should be two packages. I'd be fine with one package, after all it's logically one component...

File: package.json [hunk 7]
```diff
@@ -1,45 +1,25 @@
...
-    "typescript": "2.3.3"
-  },
-  "dependencies": {
-    "fs": "^0.0.1-security",
-    "js-yaml": "^3.8.4",
-    "jsonpath": "^0.2.11",
-    "vscode-jsonrpc": "^3.2.0"
-  }
-}
+    "name": "CommonDeps",
```
Comment (veronicagg) [inline package.json]: is it better to have a common dependencies package.json, or just have one for each package? aren't the packages themselves independent?

Comment (dsgouda) [inline package.json]: This is simply a package.json for dev dependencies (to install packages like gulp to actually run the build scripts), we will never publish this.

**Comment by dsgouda** [comment]
Will merge these changes as is, opening a new issue [here](https://github.com/Azure/azure-openapi-validator/issues/26) to ensure signing of dlls before publishing a non-preview version.

**Pull request merged**